### ISSUE TITLE
Verification of Linux Kernel Loop (rtmutex) (1)

### ISSUE THREAD

**Issue by ThomasHaas** [hernanponcedeleon/Dat3M#372]
```
	do {
		owner = *p;
	} while (cmpxchg_relaxed(p, owner,
				 owner | RT_MUTEX_HAS_WAITERS) != owner);
```
coming from [rtmutex implementation](https://github.com/torvalds/linux/blob/master/kernel/locking/rtmutex.c#L237) in the linux kernel.

_Originally posted by @hernanponcedeleon in https://github.com/hernanponcedeleon/Dat3M/issues/370#issuecomment-1400391189_

**Comment by ThomasHaas** [comment]
This loop is 1-bounded in its side effects, but we do not detect this because we don't understand that when the side effect happens (= the CAS succeeds), the loop terminates.
We will unroll this loop, but dynamic loop cutting will make sure that we only ever explore the first iteration.
This loop, however, cannot possibly deadlock(*) (even under the empty memory model), so we won't find termination issues (we still needlessly generate the deadlock encoding for this loop, but it is always unsat).

In summary, we correctly detect deadlock situations here (because they are non-existent), but we could be smarter with the unrolling.


(*) The loop cannot deadlock because whenever the thread runs in isolation it will re-read the same value twice and the CAS will succeed. It could, however, potentially livelock (very unlikely or even impossible in practice?) if another thread repeatedly intervenes by storing into memory. I don't think there is any way we can detect livelock situations.

**Comment by hernanponcedeleon** [comment]
When you say "we do not detect this", you mean statically, right?

**Comment by ThomasHaas** [comment]
Yes, we do not detect this statically but dynamically.

**Comment by hernanponcedeleon** [comment]
I am not 100% sure the program below is correct (wrt to the pointers), but I get suspicious results.

If I check both liveness and safety I get `UNKNOWN`, no matter the bound I use. The fact that the `assert(0)` is not triggered suggested that the threads do not properly terminate, but then we should detect some liveness violation, don't we?

```
#include <stdlib.h>
#include <pthread.h>
#include <assert.h>
#include <stdatomic.h>

int x;
atomic_int y;

typedef struct Task {} Task;

struct lock_s {
   _Atomic(struct Task*) owner;
};
typedef struct lock_s lock_t;

void *acquire(struct lock_s *l)
{
    Task *owner;
    do {
    	owner = atomic_load_explicit(&l->owner, __ATOMIC_RELAXED);
	} while (atomic_compare_exchange_strong_explicit(&l->owner, &owner, 1, __ATOMIC_RELAXED, __ATOMIC_RELAXED));
}

lock_t lock;

void *run(void *unused)
{
    acquire(&lock);
}

int main()
{
	pthread_t t1, t2;

	pthread_create(&t1, NULL, run, NULL);
	pthread_create(&t2, NULL, run, NULL);

	pthread_join(t1, NULL);
	pthread_join(t2, NULL);

        assert(0);

	return 0;
}
```

**Comment by ThomasHaas** [comment]
You have the condition in the loop inverted: the loop runs as long as the CAS succeeds.
I think only one thread can fail the CAS, the second one will repeatedly perform CASes inside the loop, therefore it doesn't terminate. At the same time, we do not consider this a deadlock because the thread actually writes to memory all the time.

**Comment by hernanponcedeleon** [comment]
Closing resource issues. They will remain available even if closed.
